contracts/StorageAccessible.sol

```diff
@@ -23,4 +26,61 @@ contract StorageAccessible {
         }
         return result;
     }
+
+    /**
+     * @dev Performs a delegetecall on a targetContract in the context of self.
+     * Internally reverts execution to avoid side effects (making it static). Catches revert and returns encoded result as bytes.
+     * @param targetContract Address of the contract containing the code to execute.
+     * @param calldataPayload Calldata that should be sent to the target contract (encoded method name and arguments).
+     */
+    function simulateDelegatecall(
+        address targetContract,
+        bytes memory calldataPayload
+    ) public returns (bytes memory) {
+        bytes memory innerCall = abi.encodeWithSelector(
+            SIMULATE_DELEGATECALL_INTERNAL_SELECTOR,
+            targetContract,
+            calldataPayload
+        );
+        (, bytes memory response) = address(this).call(innerCall);
+        // We try to detect if the delegatecall itself reverted (with a reason string).
+        // In this case we use this reason to revert the top level call.
+        if (isRevert(response)) {
+            revertWith(response);
+        } else {
+            return response;
+        }
+    }
+
+    /**
+     * @dev Performs a delegetecall on a targetContract in the context of self.
+     * Internally reverts execution to avoid side effects (making it static). Returns encoded result as revert message.
+     * @param targetContract Address of the contract containing the code to execute.
+     * @param calldataPayload Calldata that should be sent to the target contract (encoded method name and arguments).
+     */
+    function simulateDelegatecallInternal(
+        address targetContract,
+        bytes memory calldataPayload
+    ) public returns (bytes memory) {
+        (, bytes memory response) = targetContract.delegatecall(
+            calldataPayload
+        );
+        revertWith(response);
+    }
+
+    function isRevert(bytes memory result) public pure returns (bool) {
+        // Check if result starts with the method selector "Error(string)"
+        return
+            result.length > 4 &&
+            result[0] == 0x08 &&
+            result[1] == 0xc3 &&
+            result[2] == 0x79 &&
+            result[3] == 0xa0;
```

Comment:
One tiny suggestion (but I'm not sure its worth doing) is maybe this check should be stricter, specifically:
- First four bytes must be 0x08c379a0
- Next 32-byte word must be 0x20
- Next 32-byte work must be length
- The byte length = 4 + k*32 such that `k == ceil(len / 32) + 2`

But that might be overkill.

Reply:
Your approach is 100x better.